cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -287,16 +301,36 @@ private void handleProgramEvent(ProgramRunId programRunId, ProgramRunStatus prog
         return;
     }
     if (recordedRunRecord != null) {
+      // We need to publish the message so that the trigger subscriber can pick it up and start the trigger if
+      // necessary
       publishRecordedStatus(notification, programRunId, recordedRunRecord.getStatus());
       if (programRunStatus.isEndState()) {
         // if this is a preview run or a program within a workflow, we don't actually need to de-provision the cluster.
         // instead, we just record the state as deprovisioned without notifying the provisioner
+        // and we will emit the program status metrics for it
         boolean isInWorkflow = recordedRunRecord.getSystemArgs().containsKey(ProgramOptionConstants.WORKFLOW_NAME);
         boolean skipProvisioning =
           Boolean.parseBoolean(recordedRunRecord.getSystemArgs().get(ProgramOptionConstants.SKIP_PROVISIONING));
+
         if (isInWorkflow || skipProvisioning) {
           appMetadataStore.recordProgramDeprovisioning(programRunId, messageIdBytes);
           appMetadataStore.recordProgramDeprovisioned(programRunId, null, messageIdBytes);
+
+          Optional<ProfileId> profile = SystemArguments.getProfileIdFromArgs(programRunId.getNamespaceId(),
+                                                                             systemArguments);
+          if (!profile.isPresent()) {
+            // Each program run should be associated with a profile, so this should never happen
+            LOG.warn("Will not emit metrics for program {} without profile assigned to it, {}",
+                     programRunId, notification);
+            return;
+          }
+
+          ProfileId profileId = profile.get();
+
+          // here we get the status from the run record because when we record program deprovisioned,
+          // the status might get changed to FAILED because of the previous cluster status was PROVISIONING
+          getMetricsContextForProfile(programRunId, profileId).increment(getMetricName(recordedRunRecord.getStatus()),
```

Comment:
I think skipProvisioning is used in preview, which seems like it shouldn't get recorded. If this is the case, can you rename SKIP_PROVISIONING to IS_PREVIEW and only emit when it's not preview

Reply:
Preview has isolation from the real workspace, so any metrics emitted will not actually be written to MetricStore, so we do not need a special case for preview